NE-2192:  Openshift conditions on Gateway API status (#1871)
openshift/enhancements: enhancements/ingress/add-dns-and-loadbalancer-conditions-to-managed-gateway.md

observability they currently have with IngressController, enabling faster issue 
detection and resolution for DNS and cloud LoadBalancer failures.

### User Stories

* As a customer, I want to know when my Gateway has errors due to a DNS issue 
so that I can identify and resolve DNS-related problems quickly.
* As a customer, I want to know when my Gateway has errors due to a LoadBalancer 
issue so that I can diagnose cloud infrastructure provisioning failures.
* As a cluster administrator, I want to know when a Gateway has errors due to a 
DNS issue so that I can proactively address DNS configuration or quota problems.
* As a cluster administrator, I want to know when my Gateway has errors due to 
a LoadBalancer issue so that I can troubleshoot cloud provider integration 
issues or resource limits.

### Goals

* Add `DNSManaged` and `DNSReady` conditions to Gateway status that reflect 
whether DNS is managed and the state of DNS record provisioning
* Add `LoadBalancerReady` conditions to Gateway status that reflect whether 
LoadBalancer is managed and the state of cloud LoadBalancer service provisioning
* Implement a two-tier condition model: "Managed" conditions indicate whether 

> OpenShift should manage the resource, "Ready" conditions indicate whether it is 

Author: right, I was doing the cleanup and missed it. I will wait until you finish the review to fix other missing parts (and also add you to the reviewer list of this EP)

functioning
* Ensure conditions follow the same semantics and behavior as existing 
IngressController DNS and LoadBalancer conditions by reusing the same status 
computation logic
* Scope the feature to Gateways in the `openshift-ingress` namespace that are managed
by Openshift Gateway Class (`openshift.io/gateway-controller/v1`)
* Support cloud platforms with native LoadBalancer services (AWS, Azure, GCP, etc.)

### Non-Goals

* Supporting bare metal or on-premise deployments that do not have cloud 
LoadBalancer services
* Adding these conditions to user-managed Gateway resources outside the 
`openshift-ingress` namespace
* Modifying or changing existing IngressController condition behavior or semantics

> * Introducing custom condition types beyond DNS and LoadBalancer at this time

Reviewer: So this means we don't mark IngressController as Degraded if there are problems with the Gateway?

5. Cluster Ingress Operator detects the Gateway resource and begins reconciliation
6. Cluster Ingress Operator initiates DNS record provisioning through its own dns controller
7. Cluster Ingress Operator dns controller successfully creates DNS records and 
updates its status
8. Gateway Status Controller updates Gateway condition `LoadBalancerReady=True`
with reason "LoadBalancerProvisioned"
9. Gateway Status Controller updates Gateway condition `DNSManaged=True` with 
reason "Normal" (DNS should be managed)
10. Gateway Status Controller updates Gateway condition `DNSReady=True` with 
reason "Normal" (DNS records provisioned successfully)
11. Customer checks Gateway status and sees all four conditions are `True`, confirming the Gateway is fully operational

#### DNS Failure Flow

1. Customer creates a Gateway resource in the `openshift-ingress` namespace
2. Cluster Ingress Operator initiates DNS record provisioning through its own dns controller
3. Cluster Ingress Operator DNS controller encounters an error (e.g., invalid 
zone, quota exceeded, provider API error)
4. Cluster Ingress Operator DNS controller reports failure status in the 
DNSRecord resource

> 5. Gateway Status Controller updates Gateway condition `DNSManaged=True`

Reviewer: Is there any reason DNSManaged=False?  Maybe reserved for the future, if another DNS management system is selected, such as ExternalDNS?

* No special handling is required; the same condition logic applies

#### Standalone Clusters

This enhancement is fully relevant for standalone OpenShift clusters deployed on
cloud platforms (AWS, Azure, GCP, etc.). The conditions provide visibility into
DNS and LoadBalancer provisioning for the cluster's ingress infrastructure.

#### Single-node Deployments or MicroShift

**Single-node OpenShift (SNO):**
* Applicable only when SNO is deployed on a cloud platform with LoadBalancer support
* On bare metal SNO deployments, LoadBalancer conditions would remain in `Unknown`
or `False` state as LoadBalancer services are not available
* DNS conditions apply regardless of platform if DNS records are being managed

**MicroShift:**
* MicroShift typically does not use Gateway API or cloud LoadBalancer services
* This enhancement is not applicable to MicroShift deployments
* No impact on MicroShift resource consumption or configuration

**Resource Impact:**
* Minimal CPU/memory impact: only adds condition updates during reconciliation
* No additional controllers or processes required
* Negligible increase in etcd storage for condition status (~1KB per Gateway)

> ### Implementation Details/Notes/Constraints

Reviewer: One more implementation detail - how about making sure the status gets added to the must-gather troubleshooting document?

Author: as you've said, I think this is implementation detail so we can keep it on CIO PR. I don't know where we should add the need of must-gather here on this document tbh, or if there's any implication of adding Gateway to must-gather (like sensitive information, etc)

Author: @candita I have just confirmed this is already part of the must-gather (tested on a OCP 4.21 RC cluster), so I think this subject can be closed.

must-gather already gets all of the Gateways.

* The implementation creates a shared package that contains status computation logic
* This logic is used by both the existing IngressController status controller
and the new Gateway status controller
* Condition computation functions accept generic inputs (DNSRecord, Service,
events, config) and return conditions
* Gateway-specific wrappers convert between internal condition types
and Gateway API `metav1.Condition` types

**Resource Discovery:**
* The controller watches the following resources to trigger reconciliation:
  - Gateway resources in `openshift-ingress` namespace that uses GatewayClasses with 
  `openshift.io/gateway-controller/v1` as their `.spec.controllerName` 
  - DNSRecord resources with `gateway.networking.k8s.io/gateway-name` label
  - Service resources with `gateway.networking.k8s.io/gateway-name` label
  - DNS cluster configuration changes
  - Infrastructure cluster configuration changes

**Platform Detection:**
* Platform type is determined from the cluster Infrastructure resource
* DNS zone configuration comes from the cluster DNS resource

**Condition Lifecycle:**
* Conditions are added when a Gateway is reconciled in the `openshift-ingress` namespace

> * Conditions are updated in-place using `condutils.SetStatusCondition()` to preserve transition times

Reviewer: Why preserve transition times?  Transition time should be updated if the condition changes, at least if it changes from true to false or vice versa.

* Conditions are automatically garbage collected when the Gateway is deleted (part of Gateway status)
* Maximum of 8 total conditions are maintained per Gateway to prevent unbounded growth

**Permissions:**

> * The cluster-ingress-operator service account is granted RBAC permissions to:

Reviewer: I would like to see it say:
```suggestion
* The cluster-ingress-operator service account uses existing RBAC permissions to:
```

  - Get, List, Watch Gateway resources
  - Patch Gateway status subresource
  - Get, List, Watch DNSRecord and Service resources by label

### Risks and Mitigations

**Risk: Condition limit on Gateway resource**
Gateway resource establishes a limit of 8 conditions on its status. Istio, which is 
used as our Gateway API implementation, sets 2 conditions. Cluster Ingress Operator 

> will set additional 4 conditions. In case Istio starts adding more conditions, this 

Reviewer: nit: Just 3, not 4.

* Could use Kubernetes events to report DNS/LB status instead of conditions
* Rejected because: Events are transient and harder to query programmatically;
conditions provide persistent, queryable status

## Open Questions [optional]

<!-- TODO: This section needs to be filled in -->

1. What should the condition status be on platforms without LoadBalancer support (bare metal)?
   - Proposed: Set to `Unknown` with message indicating platform does not support LoadBalancers

2. Should these conditions be backported to older OpenShift versions?
   - To be determined based on customer demand
  
## Test Plan

**Unit Tests:**
* Test `ComputeDNSStatus` with various DNSRecord states (no zones, failed zones, successful zones, unmanaged)
* Test `ComputeLoadBalancerStatus` with various Service states (not found, pending, provisioned, failed)
* Test `ComputeGatewayAPIDNSStatus` wrapper correctly converts internal conditions to Gateway API conditions
* Test `ComputeGatewayAPILoadBalancerStatus` wrapper correctly converts internal conditions to Gateway API conditions
* Test condition computation with DNSManagementPolicy set to Managed vs Unmanaged

> * Test ObservedGeneration is correctly set on conditions

Reviewer: Maybe also:  Test transition times?

**E2E Tests:**

> * Create Gateway in `openshift-ingress` namespace and verify all four conditions become `True`

Reviewer: three, not four

  - No disruption to existing Gateways or traffic flow
* The feature is backwards compatible; older clients that don't understand 
these conditions will ignore them

**Downgrade Strategy:**
* On downgrade to a version without this enhancement:
  - The new conditions will remain in the Gateway status but will not be updated
  - The CVO does not delete status fields, so conditions will persist but become stale
  - No impact on Gateway functionality; conditions are read-only status
  - If desired, users can manually remove the conditions, but this is not required

**Version Skew:**
* Ingress operator and Gateway resources can be at different versions during upgrade
* Older ingress operator will not add/update the new conditions (graceful degradation)
* Newer ingress operator will add conditions to any Gateway in `openshift-ingress` regardless of when it was created

## Version Skew Strategy

This enhancement involves coordination between:
* Ingress operator (adds conditions to Gateway status)
* Gateway API controller (adds labels, creates services)

**Version Skew Handling:**
* No kubelet changes are involved
* No CSI, CRI, or CNI changes are involved

**Compatibility:**

> * Feature works with Gateway API v1 (both support custom conditions)

Reviewer: nit: What does the "both" refer to here?

* No breaking changes to Gateway API contract
* Follows standard Kubernetes condition conventions

## Operational Aspects of API Extensions

**API Extension Type:**
* This enhancement adds status conditions to existing Gateway resources (no new CRDs)
* No admission webhooks, conversion webhooks, aggregated API servers, or finalizers are introduced
* Status conditions are a standard Kubernetes pattern and do not require API approval for new types

**Impact on Existing SLIs:**
* Negligible impact on API throughput: condition updates happen during normal reconciliation
* No new API calls introduced; only status updates to existing Gateway resources
* Expected number of managed Gateways in `openshift-ingress` namespace: typically 1-10 per cluster

> * Condition updates are rate-limited to prevent excessive writes

Author: removed to avoid any kind of confusion

**Failure Modes:**
* If ingress operator is unavailable, conditions will not be updated, but Gateways continue to function
* Conditions are status-only and do not affect Gateway data plane functionality

**Health Indicators:**
* Monitor ingress-operator pod health and logs for errors updating Gateway conditions
* Existing ingress-operator metrics and alerts apply

**Responsible Teams:**
* Networking team (ingress): primary owner for Gateway condition logic

## Support Procedures

**Detecting Issues:**

> *Symptom: Gateway conditions show `DNSManaged=False`*

Reviewer: This is not necessarily an error condition.  Some users may choose to not have DNS be managed.

* A new `gateway-status` controller is added to the cluster-ingress-operator
* Uses the controller-runtime framework with predicates to filter relevant Gateway resources

**Condition Update Logic:**
* The ingress operator watches Gateway resources in the `openshift-ingress`
namespace that have the OpenShift Gateway Class as their `.spec.gatewayClassName` controller
* Associated DNSRecord and Service resources are discovered using the
`gateway.networking.k8s.io/gateway-name` label
* Only the first matching DNSRecord and Service in the same namespace are used
for status computation
* Conditions updated during the Gateway reconciliation:
  - DNS conditions
  - LoadBalancer conditions
* These functions wrap the existing functions used for IngressController, ensuring consistency
* The operator patches Gateway status using the Kubernetes API with the computed conditions
* Conditions include `ObservedGeneration` to track which Gateway generation the status reflects

**DNS Condition Details:**

*DNSManaged Condition:*
* Set to `False` with reason `NoDNSZones` when cluster DNS configuration has no zones
* Set to `False` with reason `UnsupportedEndpointPublishingStrategy` when the

> publishing strategy is not LoadBalancerService

Reviewer: Same question.

* Set to `False` with reason `FailedZones` when one or more zones have failed to publish
* Set to `False` with reason `UnknownZones` when zone status cannot be determined
* Set to `True` with reason `Normal` when all zones have successfully published DNS records
* Error messages include specific DNS provider errors when provisioning fails

**LoadBalancer Condition Details:**

*LoadBalancerManaged Condition:*
* Set to `False` with reason `UnsupportedEndpointPublishingStrategy` when the
publishing strategy doesn't require a managed LoadBalancer
* Set to `True` with reason `Normal` when a LoadBalancer service should be managed by OpenShift

*LoadBalancerReady Condition:*
* Set to `False` with reason `ServiceNotFound` when the associated Service
resource cannot be found
* Set to `False` with reason `LoadBalancerPending` when the service exists but
has no LoadBalancer ingress (external IP/hostname)
* Set to `False` with reason `SyncLoadBalancerFailed` when cloud provider events
indicate provisioning failures
* Set to `True` with reason `LoadBalancerProvisioned` when the service has an
external IP or hostname assigned

> * Error messages include cloud provider error details extracted from service events

Author: right, so consider that all of these conditions are a mirror of what we have on CIO. In fact, the same logic from CIO is used to generate Gateway API conditions.

I think that the LoadBalancerManaged is not just about Istio adding it, but being able to make explicit why the LoadBalancer (and the Gateway) are not ready if tere is a problem with the Load balancer controller.

That said, if we agree that the whole "lifecycle" for LoadBalancer and its conditions should be accounted directly by Istio, I think that NE-2183 makes partial sense for DNS management, but the whole Load balancer story can be dropped, wdyt?

**Code Reuse:**
* The implementation creates a shared package that contains status computation logic
* This logic is used by both the existing IngressController status controller

> and the new Gateway status controller

Reviewer 2: We have to make sure to not impact the existing ingresscontroller status logic. It's quite complex and quite touchy for our customers. We have the test coverage to catch the regression, of course. But I would prefer to limit the changes into the existing logic to very save ones.

* Condition computation functions accept generic inputs (DNSRecord, Service,
events, config) and return conditions
* Gateway-specific wrappers convert between internal condition types
and Gateway API `metav1.Condition` types

**Resource Discovery:**
* The controller watches the following resources to trigger reconciliation:

>   - Gateway resources in `openshift-ingress` namespace

Reviewer 2: With `openshift.io/gateway-controller/v1` controller, right?

  - DNSRecord resources with `gateway.networking.k8s.io/gateway-name` label
  - Service resources with `gateway.networking.k8s.io/gateway-name` label
  - DNS cluster configuration changes
  - Infrastructure cluster configuration changes

**Platform Detection:**
* Platform type is determined from the cluster Infrastructure resource
* DNS zone configuration comes from the cluster DNS resource
* The DNSManagementPolicy field on IngressController-like configuration determines whether DNS is managed

**Condition Lifecycle:**
* Conditions are added when a Gateway is reconciled in the `openshift-ingress` namespace
* Conditions are updated in-place using `condutils.SetStatusCondition()` to preserve transition times
* Conditions are automatically garbage collected when the Gateway is deleted (part of Gateway status)
* Maximum of 8 total conditions are maintained per Gateway to prevent unbounded growth

**Permissions:**
* The cluster-ingress-operator service account is granted additional RBAC permissions to:

>   - Get, List, Watch Gateway resources

Reviewer 2: CIO already has [all the verbs access for Gateway resources](https://github.com/openshift/cluster-ingress-operator/blob/0cac97af560a68a5f5c0390a529726eb43556d51/manifests/00-cluster-role.yaml#L161-L169).

**UX Review:**
* Condition messages should be clear and actionable for both administrators and developers
* Error messages should include enough detail to enable self-service troubleshooting

### Drawbacks

**Additional API surface:**
* Adding conditions increases the Gateway status size, though the impact is
minimal (~1KB per Gateway)

**Maintenance burden:**
* Requires ongoing maintenance to keep condition logic in sync with DNS operator and cloud provider behavior
* May need updates when DNS operator or cloud provider integrations change

**Potential confusion:**
* Users may expect these conditions to appear on all Gateway resources, not just
those in `openshift-ingress`
* Clear documentation will be needed to explain the scoping

**Not applicable to all environments:**
* The LoadBalancer condition is only meaningful on cloud platforms or platforms with
`LoadBalancer` support.
* Users on bare metal may see persistent `False` or `Unknown` status which could
be confusing

## Alternatives (Not Implemented)

**Alternative 1: Use Gateway API standard conditions only**
* The Gateway API specification defines standard conditions like `Accepted` and `Programmed`

> * We could map DNS/LB status into these existing conditions rather than adding new ones

Reviewer 2: Agree about the DNS part but not for the LB one which is managed by Gateway implementation (==Istio).

* Rejected because: Standard conditions don't provide the granularity needed to
distinguish DNS vs. LoadBalancer issues

**Alternative 2: Add conditions to all Gateway resources cluster-wide**
* Could add these conditions to any Gateway resource, not just those in `openshift-ingress`
* Rejected because: User-managed Gateways may not follow the same DNS/LB
provisioning patterns, and this would create inconsistent behavior

**Alternative 3: Create separate status CRD for infrastructure details**
* Could create a new CRD (e.g., `GatewayInfrastructureStatus`) to hold DNS/LB information
* Rejected because: Adds unnecessary complexity; conditions are the standard
Kubernetes pattern for status reporting

**Alternative 4: Emit events instead of conditions**
* Could use Kubernetes events to report DNS/LB status instead of conditions
* Rejected because: Events are transient and harder to query programmatically;
conditions provide persistent, queryable status

> ## Open Questions [optional]

Reviewer 2: What about conflicting hostnames? With the current DNS implementation from CIO, DNS resource record will have multiple entries for the same hostname round robin between gateway controllers.

Reviewer 3: > With the current DNS implementation from CIO, DNS resource record will have multiple entries for the same hostname round robin between gateway controllers.

We only publish a DNS record for one of two conflicting gateways.  Unfortunately, the exact behavior varies by platform: on AWS, the second wins; on GCP, the first wins.  I don't know what happens on Azure, IBM Cloud, or other platforms.  That's all the more reason we need https://github.com/openshift/cluster-ingress-operator/pull/1229.

observability they currently have with IngressController, enabling faster issue 
detection and resolution for DNS and cloud LoadBalancer failures.

### User Stories

* As a cluster administrator, I want to know which specific Gateway listeners have
DNS issues so that I can proactively address DNS configuration or quota problems
for specific hostnames.
* As a cluster administrator, I want to know when my Gateway has errors due to
a LoadBalancer issue so that I can troubleshoot cloud provider integration
issues or resource limits.

### Goals

* Add `DNSReady` condition to each Gateway listener status (in `status.listeners[].conditions[]`)
that reflects the state of DNS record provisioning for that specific listener
* Add `LoadBalancerReady` condition to Gateway status (in `status.conditions[]`)
that reflects the state of cloud LoadBalancer service provisioning
* Ensure conditions follow the same semantics and behavior as existing
IngressController DNS and LoadBalancer conditions by reusing the same status
computation logic
* Scope the feature to Gateways in the `openshift-ingress` namespace that are managed

> by Openshift Gateway Class (`openshift.io/gateway-controller/v1`)

Reviewer 3: ```suggestion
by OpenShift gateway controller name (`openshift.io/gateway-controller/v1`)
```

    is functioning correctly

* **Gateway-level conditions** (in `status.conditions[]`):
  - `LoadBalancerReady`: Indicates whether the cloud LoadBalancer service for
    the entire Gateway is functioning correctly

These conditions will be managed by a new gateway-status controller in the
cluster-ingress-operator. The placement reflects the architecture where:
- Each listener with a unique hostname gets its own DNSRecord resource
- The entire Gateway shares a single LoadBalancer Service resource

The `DNSReady` and `LoadBalancerReady` conditions indicate whether the managed
resources (DNS records and LoadBalancer service) are functioning correctly.

The implementation reuses the existing status computation logic from the
IngressController by:
1. Creating a shared `pkg/resources/status` package with condition computation
functions
2. Refactoring existing IngressController status code to use this shared package
3. Creating Gateway-specific wrapper functions that call the shared logic and
convert to Gateway API condition types

The gateway-status controller will:
1. Watch Gateway resources in `openshift-ingress` namespace that specify the

> OpenShift Gateway Class

Reviewer 3: Writing "the" OpenShift GatewayClass implies there can only be one, but there could be multiple.  
```suggestion
1. Watch Gateway resources in `openshift-ingress` namespace that specify a
GatewayClass belonging to OpenShift.
```
or more explicitly
```suggestion
1. Watch Gateway resources in `openshift-ingress` namespace that specify any
GatewayClass that specifies OpenShift's gateway controller name.
```

> OpenShift Gateway Class

Reviewer 3: Writing "the" OpenShift GatewayClass implies there can only be one, but there could be multiple.  
```suggestion
1. Watch Gateway resources in `openshift-ingress` namespace that specify a
GatewayClass belonging to OpenShift.
```
or more explicitly
```suggestion
1. Watch Gateway resources in `openshift-ingress` namespace that specify any
GatewayClass that specifies OpenShift's gateway controller name.
```

   - Find the corresponding DNSRecord by matching the listener's hostname with
     the DNSRecord's `.spec.dnsName` field
   - Compute DNS conditions using the shared status logic (same as IngressController)
   - Update the listener's status in `status.listeners[].conditions[]`
4. For the Gateway's LoadBalancer service:
   - Fetch the single Service resource for the Gateway
   - Compute LoadBalancer condition using the shared status logic
   - Update the Gateway's status in `status.conditions[]`
5. Fetch cluster DNS and Infrastructure configuration
6. Patch Gateway status with the computed conditions including ObservedGeneration

This is a purely additive change that does not modify existing Gateway behavior 
or APIs. The conditions provide read-only status information to users and 
monitoring systems.

### Workflow Description

**Cluster Administrator** is a user who manages OpenShift cluster infrastructure
and troubleshoots platform issues.

**Customer/Developer** is a user who deploys applications and monitors
application ingress health.

**Ingress Operator** is the OpenShift operator responsible for managing ingress
resources including Gateways and IngressControllers.

> **DNS Operator** is responsible for managing DNS records for cluster ingress.

Reviewer 3: The DNS Operator is not relevant to this enhancement.  I notice it is referenced several times later on in this EP; all of these references are bogus.

**Cloud Provider API** is the underlying cloud infrastructure that provisions
LoadBalancer services.

#### Normal Flow (Success Case)

1. Customer creates or updates a Gateway resource with a single or multiple listeners in the `openshift-ingress` namespace
2. Gateway API controller (Istio) creates LoadBalancer service for the Gateway
3. Cloud Provider API provisions the LoadBalancer successfully
4. LoadBalancer service status is updated with external IP/hostname

> 5. Cluster Ingress Operator detects the Gateway resource and begins reconciliation

Reviewer 3: I got confused around this step and had to reread the steps a few times to understand them.  Steps 5 and 6 refer to the Cluster Ingress Operator _gateway-service-dns_ controller's actions, right?  And Steps 6 and 7 seem to refer to the dns controller's actions.  

It is a bit confusing how these steps blur the lines between controllers.  It is also confusing that these steps are written as if the gateway-status controller would wait until CCM and the dns controller finished their work, which isn't probably what will happen.  I expect it to go something more like this:

1. Customer creates the Gateway.  
2. Istio observes the Gateway's creation and creates the Service.  At the same time, CIO's gateway-status controller observes the Gateway's creation and sets `LoadBalancerReady` to `False` (because no Service exists yet) and `DNSReady` to `False` (because no DNSRecord CRs exist).  
3. CCM's service controller observes the Service's creation from Istio, provisions the SLB, and updates the Service's status (this step is currently enumerated as steps 3 and 4).  At the same time, CIO's gateway-status controller observes the Service's creation from Istio but continues to report `LoadBalancerReady` as `False` (because no SLB is provisioned yet) and `DNSReady` to `False`.  
4. CIO's gateway-service-dns controller observes the Service update from CCM and creates DNSRecord CRs (currently enumerated as steps 5 and 6).  At the same time, CIO's gateway-status controller observes the Service update from CCM and sets `LoadBalancerReady` to `True` and `DNSReady` to `False`.  
5. CIO's dns controller observes the DNSRecord CRs' creation, creates the DNS records, and updates the DNSRecord CRs' statuses (currently steps 6 and 7).  
6. CIO's gateway-status controller observes the DNSRecord CR updates and updates the Gateway's status by setting `LoadBalancerReady` to `True` and `DNSReady` to `True` (currently steps 8 and 9).  (There is a possibility with multiple DNSRecord CRs that not all report success at the same time, in which case the gateway-status controller might run multiple times and initially set `DNSReady` to `False` until all DNSRecord CRs are published whereupon the controller would set DNSReady` to `True`.)

Do I have that correct?  Is explaining it in terms of "controller X observes event Y and does action Z" make it clearer?

Author: yes, but I think writing all of the controllers actuating at that moment like this is a bit massive to read (at least for me)

Maybe should this be just "reduced" into Gateway created, Istio creates a service, Gateway Status Controller waits for the Service and DNS to be ready to set the right conditions?

I think that the full workflow containing all of the steps that a condition can get until it is successful very overwhelming.

> 6. Cluster Ingress Operator initiates DNS record provisioning through its own dns controller,

Author: fixed

(e.g., invalid zone, quota exceeded, provider API error)
4. Cluster Ingress Operator DNS controller reports failure status in the affected
DNSRecord resource
5. Gateway Status Controller updates the affected listener's conditions in
`status.listeners[].conditions[]`:
   - Sets `DNSReady=False` with reason `FailedZones` and detailed error message from DNS provider
6. Other listeners with successful DNS records show `DNSReady=True`
7. Cluster Administrator reviews Gateway status and identifies which specific listener
has DNS issues by checking each listener's conditions
8. Cluster Administrator resolves the DNS issue (e.g., increases quota,
fixes zone configuration)
9. DNS Operator retries and successfully provisions the DNS record
10. Gateway Status Controller updates the affected listener's condition `DNSReady=True` with
reason "NoFailedZones"

#### LoadBalancer Failure Flow

1. Customer creates a Gateway resource in the `openshift-ingress` namespace
2. Gateway API controller creates LoadBalancer service for the Gateway
3. Cloud Provider API fails to provision LoadBalancer (e.g., quota exceeded,
subnet full, invalid configuration)

> 4. LoadBalancer service remains in Pending state with event describing the error

Reviewer 3: Fair enough.  I didn't mean to say you needed to change the fundamental design.  If the issue of inaccurate status due to the transient nature of events comes up again in the future, we can fix it at that time, for both the IngressController status and Gateway status at the same time.

```

### API Extensions

No API extension will be made, just conditions being added to the already existing 
Gateway resource.

### Topology Considerations

#### Hypershift / Hosted Control Planes

This enhancement applies to Hypershift environments with the following considerations:

* The ingress operator runs in the management cluster but manages Gateway resources for the hosted cluster
* DNS and LoadBalancer provisioning happens in the context of the hosted cluster's infrastructure
* Conditions will reflect the status of infrastructure provisioned for the hosted cluster
* No special handling is required; the same condition logic applies

#### Standalone Clusters

This enhancement is fully relevant for standalone OpenShift clusters deployed on
cloud platforms (AWS, Azure, GCP, etc.). The conditions provide visibility into
DNS and LoadBalancer provisioning for the cluster's ingress infrastructure.

For on-premises environments, the same behavior existing on Ingress Controllers is replicated
to Gateway API status: the `LoadBalancerReady` will reflect the existence of a Load balancer

> controller on the environment, while `DNSReady` will reflect the DNS record provisioning status.

Author: added

    port: 80
    protocol: HTTP
  - name: stage-https
    hostname: "*.stage.example.com"
    port: 443
    protocol: HTTPS
  - name: prod-https
    hostname: "*.prod.example.com"
    port: 443
    protocol: HTTPS
status:
  # Gateway-level conditions (LoadBalancer status)
  conditions:
  - type: LoadBalancerReady
    status: "True"
    reason: LoadBalancerProvisioned
    message: "The LoadBalancer service is provisioned"
    observedGeneration: 1
    lastTransitionTime: "2025-01-12T10:00:00Z"

  # Listener-level conditions (DNS status per listener)
  listeners:
  - name: stage-http
    conditions:
    - type: DNSReady
      status: "True"
      reason: NoFailedZones
      message: "The record is provisioned in all reported zones."
      observedGeneration: 1
      lastTransitionTime: "2025-01-12T10:00:00Z"

  - name: stage-https
    conditions:
    - type: DNSReady
      status: "True"
      reason: NoFailedZones
      message: "The record is provisioned in all reported zones."
      observedGeneration: 1
      lastTransitionTime: "2025-01-12T10:00:00Z"

  - name: prod-https
    conditions:
    - type: DNSReady
      status: "False"
      reason: FailedZones

>       message: "The record failed to provision in some zones: [prod.example.com]"

Reviewer 3: Actually, you do say here that the status message should have a "detailed error message": https://github.com/openshift/enhancements/blob/1a6725a5dd299963a213144880dcac7d01afb0ae/enhancements/ingress/add-dns-and-loadbalancer-conditions-to-managed-gateway.md#L186

But again, this can be handled in a follow-up.

to prevent unbounded growth

**Permissions:**
* The cluster-ingress-operator service account uses existing RBAC permissions to:
  - Get, List, Watch Gateway resources
  - Get, List, Watch DNSRecord and Service resources by label
* The cluster-ingress-operator service account needs new RBAC permissions to:
  - Update and Patch `gateways.gateway.networking.k8s.io` status subresource

### Risks and Mitigations

**Risk: Condition limit on Gateway and Listener resources**
Gateway API establishes a limit of 8 conditions on both the Gateway status and each
listener status. Istio, which is used as our Gateway API implementation, sets 2
gateway-level conditions and also sets listener-level conditions. Cluster Ingress
Operator will set 1 additional gateway-level condition (`LoadBalancerReady`) and
1 listener-level condition per listener (`DNSReady`). If Istio starts adding more
conditions, this can be a problem.

* Mitigation:
  - The implementation intentionally does NOT add `LoadBalancerManaged` or `DNSManaged`
    conditions to Gateway resources to conserve the limited condition slots. These
    "managed" conditions would always be `True` for Gateways in `openshift-ingress`

>     namespace and provide limited value compared to the "ready" conditions.

Reviewer 3: The statement in the EP that these conditions "would always be `True`" isn't accurate; using the existing logic that we use for ingresscontrollers, a `DNSManaged` status condition would have status `False` if the listener specified a host name outside the cluster's base domain on AWS or GCP.  However, we did agree that `DNSManaged` wasn't strictly necessary, so I am fine with omitting it.

Author: ok just removed the last phrase

  - An e2e test should be added to assert that the total number of conditions
    at both gateway and listener levels is consistent with the current Istio and Cluster
    Ingress Operator conditions. In case of change, the test should fail and engineering
    should verify if Istio is adding additional conditions.

**Risk: Exposure of sensitive information in condition messages**
* Mitigation: Sanitize error messages to remove any credentials or sensitive
data before including in conditions

**Risk: Compatibility with future Gateway API versions**
* Mitigation: Use well-defined condition types that align with Kubernetes

> conventions and Gateway API patterns

Reviewer 3: I'm not sure what the risk is here, or how we mitigate it.  Is the risk that upstream defines a conflicting status condition that uses the same "DNSReady" or "LoadBalancerReady" type?

Author: probably halucination, I have missed it, removing

**Security Review:**
* No new authentication or authorization mechanisms introduced
* Conditions are read-only status information
* Standard RBAC applies to Gateway resources
* Cluster Ingress Operator cluster role needs a new permission to `UPDATE` and `PATCH`
Gateway status subresource.

**UX Review:**
* Condition messages should be clear and actionable for both administrators and developers

> * Error messages should include enough detail to enable self-service troubleshooting

Author: it came from the original template when it says about "UX"...removed it

**Maintenance burden:**
* Requires ongoing maintenance to keep condition logic in sync with DNS operator and cloud provider behavior
* May need updates when DNS operator or cloud provider integrations change

**Potential confusion:**
* Users may expect these conditions to appear on all Gateway resources, not just
those in `openshift-ingress`
* Users need to understand the difference between gateway-level conditions (LoadBalancer)
and listener-level conditions (DNS)
* Clear documentation will be needed to explain the scoping and condition placement

**Not applicable to all environments:**
* The LoadBalancer condition is only meaningful on cloud platforms or platforms with
`LoadBalancer` support.
* Users on bare metal may see persistent `False` or `Unknown` status which could
be confusing

## Alternatives (Not Implemented)

**Alternative 1: Use Gateway API standard conditions only**
* The Gateway API specification defines standard conditions like `Accepted` and `Programmed`
* We could map DNS/LB status into these existing conditions rather than adding new ones
* Rejected because: Standard conditions don't provide the granularity needed to

> distinguish DNS vs. LoadBalancer issues

Reviewer 3: More importantly, Istio already sets the standard conditions, and the proposed gateway-status controller would conflict with Istio if they both tried to set the same conditions.

**Alternative 2: Add conditions to all Gateway resources cluster-wide**
* Could add these conditions to any Gateway resource, not just those in `openshift-ingress`
* Rejected because: User-managed Gateways may not follow the same DNS/LB

> provisioning patterns, and this would create inconsistent behavior

Author: agreed, but during our initial discussion for it we decided to go first with openshift-ingress because it is a namespace managed by admin and also because it wouldn't make much sense to add conditions if we didn't created the DNSRecords for gateways out of openshift-ingress namespaces.

That said, I agree we should improve it, but I would like to do it as a followup of this EP.

Kubernetes pattern for status reporting

**Alternative 4: Emit events instead of conditions**
* Could use Kubernetes events to report DNS/LB status instead of conditions
* Rejected because: Events are transient and harder to query programmatically;
conditions provide persistent, queryable status

**Alternative 5: Put DNS conditions at Gateway level instead of listener level**
* Could add DNS conditions to `status.conditions[]` (gateway-level) and aggregate
status across all DNSRecords
* Rejected because:
  - Each listener with a unique hostname gets its own DNSRecord resource
  - A single gateway-level DNS condition cannot accurately represent the state when
    some listeners have working DNS and others don't
  - Users need to know which specific listener has DNS issues for troubleshooting
  - Gateway API supports listener-level conditions specifically for this use case
  - The current implementation in cluster-ingress-operator creates per-hostname DNSRecords,
    making listener-level status the natural fit
  - Multiple listeners can share the same hostname (and thus the same DNSRecord), so
    listener-level conditions allow showing the same DNS status for all listeners

>     sharing a hostname

Reviewer 3: I would expect `DNSReady` on `Gateway.status.conditions[]` to have status `True` if and only if all of the corresponding DNSRecord CRs reported `Published` with status `True`.  Then your `kubectl wait --for=condition=DNSReady gateway/foo` command would wait for all DNS records to have been published.

   balancer provisioned by the Gateway API controller (Istio).
  
## Test Plan

**Unit Tests:**
* Test `ComputeDNSStatus` with various DNSRecord states (no zones, failed zones, successful zones)
* Test `ComputeLoadBalancerStatus` with various Service states (not found, pending, provisioned, failed)
* Test `ComputeGatewayAPIDNSStatus` wrapper correctly converts internal conditions to Gateway API listener conditions
* Test `ComputeGatewayAPILoadBalancerStatus` wrapper correctly converts internal conditions to Gateway API gateway-level conditions
* Test ObservedGeneration is correctly set on conditions
* Test if the transition times are correctly set on conditions
* Test matching DNSRecords to listeners by hostname
* Test handling of listeners without hostnames (no DNS conditions should be added)
* Test handling of multiple listeners with the same hostname (should map to same DNSRecord)

**E2E Tests:**
* Create Gateway in `openshift-ingress` namespace with listeners and verify:
  - Gateway-level condition `LoadBalancerReady=True` in `status.conditions[]`
  - Each listener has `DNSReady=True` in `status.listeners[].conditions[]`

> * On the same test, verify that changing Gateway generation will bump the conditions `observedGeneration`

Reviewer 3: What does "Gateway generation" mean here?

* On the same test, verify that the Gateway API controller (Istio) reconciliation
does not replace OpenShift conditions
* On the same test, verify the condition count is consistent with Istio and OpenShift
added conditions (both at gateway and listener levels)
* Create Gateway out of `openshift-ingress` and verify that no OpenShift condition is added

> * Create Gateway with wrong DNS Domain and verify that listener DNS conditions reflect the failure

Reviewer 3: What does "wrong" mean here?

  - Existing Gateway resources in `openshift-ingress` namespace will have 
  conditions added during next reconciliation
  - No user action required
  - No disruption to existing Gateways or traffic flow
* The feature is backwards compatible; older clients that don't understand 
these conditions will ignore them

**Downgrade Strategy:**
* On downgrade to a version without this enhancement:
  - The new conditions will remain in the Gateway status but will not be updated
  - The CVO does not delete status fields, so conditions will persist but become stale
  - No impact on Gateway functionality; conditions are read-only status
  - If desired, users can manually remove the conditions, but this is not required

**Version Skew:**
* Ingress operator and Gateway resources can be at different versions during upgrade
* Older ingress operator will not add/update the new conditions (graceful degradation)
* Newer ingress operator will add conditions to any Gateway in `openshift-ingress` regardless of when it was created

## Version Skew Strategy

This enhancement involves coordination between:
* Ingress operator (adds conditions to Gateway status)

> * Gateway API controller (adds labels, creates services)

Reviewer 3: "Gateway API controller" refers to Istio?  It might be clearer to say that explicitly.  (Presumably the Service is the one for the Gateway, and the label is the `gateway.networking.k8s.io/gateway-name` label.)

**Version Skew Handling:**
* No kubelet changes are involved
* No CSI, CRI, or CNI changes are involved

**Compatibility:**
* Feature works with Gateway API v1
* No breaking changes to Gateway API contract
* Follows standard Kubernetes condition conventions

## Operational Aspects of API Extensions

**API Extension Type:**
* This enhancement adds status conditions to existing Gateway resources (no new CRDs)
* Conditions are added at two levels:
  - Gateway-level conditions in `status.conditions[]` for LoadBalancer status
  - Listener-level conditions in `status.listeners[].conditions[]` for DNS status
* No admission webhooks, conversion webhooks, aggregated API servers, or finalizers are introduced
* Status conditions are a standard Kubernetes and Gateway API pattern and do not require API approval for new types

**Impact on Existing SLIs:**
* Negligible impact on API throughput: condition updates happen during normal reconciliation

> * No new API calls introduced; only status updates to existing Gateway resources

Reviewer 3: "No new API calls introduced" seems like an overly broad claim (status updates aren't API calls? What about the events monitoring?).  However, I agree that the impact should be negligible (barring any unforeseen issues, such as some conflict with another controller).

> * Expected number of managed Gateways in `openshift-ingress` namespace: typically 1-10 per cluster

Reviewer 3: Is this based on actual data?  I don't know that it matters, and it is a little surprising to see this topic raised.

* For DNS conditions: Check that listeners have hostnames specified in `.spec.listeners[].hostname`
  (DNS conditions are only added to listeners with hostnames)
* For DNS conditions: If a listener previously had a hostname but it was removed, the DNSReady condition
  will be removed from that listener's status
* For LoadBalancer conditions: Check in `status.conditions[]` at the Gateway level
* For DNS conditions: Check in `status.listeners[].conditions[]` for each listener

**Disabling the Feature:**
* This feature cannot be easily disabled as it's integrated into ingress operator reconciliation
* Conditions are harmless status fields and do not need to be disabled
* If needed, users can ignore these conditions; they do not affect Gateway functionality

**Graceful Failure:**
* Condition updates are best-effort; failures to update conditions do not block Gateway provisioning
* If condition updates fail, ingress operator logs errors but continues reconciliation
* Stale conditions do not prevent Gateways from functioning correctly

**Bug Reports:**

> 	* For DNS-related issues: open an OCP bug in the Networking / DNS component

Reviewer 3: "Networking / router" should be used for DNS-related issues too; we use "Networking / router" for issues related to cluster-ingress-operator.  That said, we'll see the report if even it has "Networking / DNS" instead.

Author: fixed

	* For LoadBalancer issues: open an OCP bug in the Cloud Compute / <cloud provider> component
	* For condition update issues: open an OCP bug in the Networking / router component

## Infrastructure Needed [optional]

No new infrastructure required. This enhancement uses existing OpenShift components:
* Ingress operator (existing)

> * DNS operator (existing)

Reviewer 3: This enhancement does not use the DNS operator.

> * Cloud controller manager / LoadBalancer services (existing)

Reviewer 3: Maybe worth mentioning MetalLB, and vSphere for testing it.

**MicroShift:**
* This enhancement is not applicable to MicroShift deployments
* No impact on MicroShift resource consumption or configuration

**Resource Impact:**
* Minimal CPU/memory impact: only adds condition updates during reconciliation
* A new `gateway-status` controller is created on existing Cluster Ingress Operator
* Negligible increase in etcd storage for condition status (~1KB per Gateway)

#### OpenShift Kubernetes Engine

No special considerations for OpenShift Kubernetes Engine.

### Implementation Details/Notes/Constraints

**Architecture Overview:**

This enhancement maps infrastructure resources to Gateway status as follows:

```
Gateway (1)
├── Spec
│   └── Listeners (N)
│       ├── Listener[0] (hostname: *.stage.example.com)
│       ├── Listener[1] (hostname: *.prod.example.com)
│       └── Listener[2] (hostname: *.stage.example.com)  # shares hostname with Listener[0]
│
└── Infrastructure Resources
    ├── Service (1) ────────────────────► status.conditions[]: LoadBalancerReady
    │   └── LoadBalancer (cloud)
    │
    └── DNSRecords (N unique hostnames)
        ├── DNSRecord[0] (*.stage.example.com) ──► status.listeners[0].conditions[]: DNSManaged, DNSReady

>         │                                        └► status.listeners[2].conditions[]: DNSManaged, DNSReady

Author: array number here is very clear for me, on how they are related with the same array as above. I am worried that using names implies we are on a map and not on an ordered array.

      status: "True"
      reason: Normal
      message: "...."
      observedGeneration: 1
      lastTransitionTime: "2025-01-12T10:00:00Z"
    - type: DNSReady
      status: "False"
      reason: FailedZones
      message: "Failed to publish DNS records: quota exceeded in zone prod.example.com"
      observedGeneration: 1
      lastTransitionTime: "2025-01-12T10:00:00Z"
```

Note: In this example, `stage-http` and `stage-https` share the same hostname, so they
have identical DNS conditions (both map to the same DNSRecord). The `prod-https` listener
has a different hostname and shows a DNS failure, while the LoadBalancer is working for
the entire Gateway.

**Controller Architecture:**
* A new `gateway-status` controller is added to the cluster-ingress-operator
* Uses the controller-runtime framework with predicates to filter relevant Gateway resources

**Condition Update Logic:**
* The ingress operator watches Gateway resources in the `openshift-ingress`
namespace that have the OpenShift Gateway Class as their `.spec.gatewayClassName` controller
* Associated DNSRecord and Service resources are discovered using the
`gateway.networking.k8s.io/gateway-name` label

> * Only the first matching Service in the same namespace is used

Author: added
